File: Source/Terrain/Geology.cpp
/* Copyright 2010 Kenneth Ferland

This file is part of Khazad.

Khazad is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Khazad is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Khazad.  If not, see <http://www.gnu.org/licenses/> */


#include <Geology.h>

#include <DataManager.h>
#include <Random.h>
#include <Coordinates.h>
#include <Cell.h>
#include <math.h>

#include <boost/functional/hash.hpp>


Geology::Geology()
{

}

Geology::~Geology()
{

}

bool Geology::Init(const char* SeedString)
{
    boost::hash< const char* > SeedHash;

    Ogre::Plane RockLayer = Ogre::Plane(Ogre::Vector3(0, 0, 1), 1);
    int RockType = 42;

    MasterSeed = 42; //SeedHash(SeedString);

    RandGenerator = new RandomNumberGenerator();
    RandGenerator->Init();

    NoiseGenerator = new Noise();

    LayerBoundries.push_back(make_pair(RockLayer, RockType));

    return true;
}

void Geology::GenerateWorldHeightMap(int32_t X, int32_t Y)
{
    WorldSizeX = X;
    WorldSizeY = Y;

    WorldHeight = new float*[X + 1];

    for (int32_t x = 0; x < X + 1; x++)
    {
        WorldHeight[x] = new float[Y + 1];

        for (int32_t y = 0; y < Y + 1; y++)
        {
            WorldHeight[x][y] = (NoiseGenerator->simplex_noise(2, x, y, 0) * 4);
            //WorldHeight[x][y] = x + y;
        }
    }
}

void Geology::GenerateCellEdge(float X, float Y, float heightScale, float Roughness)
{
    boost::hash< std::pair< float, float > > CoordinateHash;
    int32_t Hash = CoordinateHash(std::make_pair(Y, X));

    RandGenerator->Seed(MasterSeed);
    int32_t FinalSeed = RandGenerator->Roll(0, 0x7fffffff) ^ Hash;

    RandGenerator->Seed(FinalSeed);


    Edge[0] = X;     Edge[CELLEDGESIZE] = Y;

	float ratio = (float) pow (2.0, -Roughness);
	float scale = heightScale * ratio;

    /* Seed the endpoints of the array. To enable seamless wrapping,
       the endpoints need to be the same point. */
    uint8_t stride = CELLEDGESIZE / 2;

    while (stride)
    {
		for (int8_t i = stride; i < CELLEDGESIZE; i += stride)
		{
			Edge[i] = scale * RandGenerator->Roll(-1.0f, 1.0f) + ((Edge[i - stride] + Edge[i + stride]) * .5f);

			/* reduce random number range */
			scale *= ratio;

			i += stride;
		}
		stride >>= 1;
    }
}

void Geology::GenerateCellHeight(int32_t X, int32_t Y, float heightScale, float Roughness)
{
    uint8_t size = CELLEDGESIZE + 1;

    // Mark edges as Seeded to prevent them from being overwritten
    for (uint16_t x = 0; x < size; x++)
    {
        for (uint16_t y = 0; y < size; y++)
        {
            Seeded[x][y] = false;
            //Height[x][y] = 0.0f;
        }
    }

    // Edges Initialized to anchor values to create contiguousness
    GenerateCellEdge(WorldHeight[X][Y], WorldHeight[X + 1][Y], heightScale, Roughness);

    for (uint16_t x = 0; x < size; x++)
    {
        Height[x][0] = Edge[x];
        Seeded[x][0] = true;
    }

    GenerateCellEdge(WorldHeight[X][Y + 1], WorldHeight[X + 1][Y + 1], heightScale, Roughness);

    for (uint16_t x = 0; x < size; x++)
    {
        Height[x][CELLEDGESIZE] = Edge[x];
        Seeded[x][CELLEDGESIZE] = true;
    }

    GenerateCellEdge(WorldHeight[X][Y], WorldHeight[X][Y + 1], heightScale, Roughness);

    for (uint16_t y = 0; y < size; y++)
    {
        Height[0][y] = Edge[y];
        Seeded[0][y] = true;
    }

    GenerateCellEdge(WorldHeight[X + 1][Y], WorldHeight[X + 1][Y + 1], heightScale, Roughness);

    for (uint16_t y = 0; y < size; y++)
    {
        Height[CELLEDGESIZE][y] = Edge[y];
        Seeded[CELLEDGESIZE][y] = true;
    }

    /* initialize random number generator */

    boost::hash< std::pair< int32_t, int32_t > > CoordinateHash;
    int32_t Hash = CoordinateHash(std::make_pair(X, Y));

    RandGenerator->Seed(MasterSeed);
    int32_t FinalSeed = RandGenerator->Roll(0, 0x7fffffff) ^ Hash;

    RandGenerator->Seed(FinalSeed);


	/* Set up our roughness constants.
	   Random numbers are always generated in the range 0.0 to 1.0.
	   'scale' is multiplied by the randum number.
	   'ratio' is multiplied by 'scale' after each iteration
	   to effectively reduce the randum number range.
	   */

    /* Seed the first four values. For example, in a 4x4 array, we
       would initialize the data points indicated by '*':

           *   .   .   .   *

           .   .   .   .   .

           .   .   .   .   .

           .   .   .   .   .

           *   .   .   .   *

       In terms of the "diamond-square" algorithm, this gives us
       "squares".

       We want the four corners of the array to have the same
       point. This will allow us to tile the arrays next to each other
       such that they join seemlessly. */

    uint8_t stride = CELLEDGESIZE / 2;

    Height[0][0] = WorldHeight[X][Y];
    Seeded[0][0] = true;

    Height[CELLEDGESIZE][0] = WorldHeight[X + 1][Y];
    Seeded[CELLEDGESIZE][0] = true;

    Height[0][CELLEDGESIZE] = WorldHeight[X][Y + 1];
    Seeded[0][CELLEDGESIZE] = true;

    Height[CELLEDGESIZE][CELLEDGESIZE] = WorldHeight[X + 1][Y + 1];
    Seeded[CELLEDGESIZE][CELLEDGESIZE] = true;

    CellTopZ = max(max(Height[0][0], Height[CELLEDGESIZE][0]), max(Height[0][CELLEDGESIZE], Height[CELLEDGESIZE][CELLEDGESIZE]));
    CellBottomZ = min(min(Height[0][0], Height[CELLEDGESIZE][0]), min(Height[0][CELLEDGESIZE], Height[CELLEDGESIZE][CELLEDGESIZE]));


    // Set initial Fractal value range (scale) and the rate of decrese (ratio)
	float ratio = (float) pow (2.0, -Roughness);
	float scale = heightScale * ratio;

    /* Now we add ever-increasing detail based on the "diamond" seeded
       values. We loop over stride, which gets cut in half at the
       bottom of the loop. Since it's an int, eventually division by 2
       will produce a zero result, terminating the loop. */

    while (stride)
    {
		/* Take the existing "square" data and produce "diamond"
		   data. On the first pass through with a 4x4 matrix, the
		   existing data is shown as "X"s, and we need to generate the
	       "*" now:

               X   .   .   .   X

               .   .   .   .   .

               .   .   *   .   .

               .   .   .   .   .

               X   .   .   .   X

	      It doesn't look like diamonds. What it actually is, for the
	      first pass, is the corners of four diamonds meeting at the
	      center of the array. */
		for (uint16_t x = stride; x < CELLEDGESIZE; x += stride)
		{
			for (uint16_t y = stride; y < CELLEDGESIZE; y += stride)
			{
			    if (!Seeded[x][y])
			    {
			        float Average = (Height[x - stride][y - stride] + Height[x - stride][y + stride] + Height[x + stride][y - stride] + Height[x + stride][y + stride]) * 0.25;
                    float RandomFactor = scale * RandGenerator->Roll(-1.0f, 1.0f);

                    Height[x][y] = RandomFactor + Average;

                    if (Height[x][y] > CellTopZ)
                    {
                        CellTopZ = Height[x][y];
                    }
                    if (Height[x][y] < CellBottomZ)
                    {
                        CellBottomZ = Height[x][y];
                    }
			    }

				y += stride;
			}
			x += stride;
		}

		/* Take the existing "diamond" data and make it into
	       "squares". Back to our 4X4 example: The first time we
	       encounter this code, the existing values are represented by
	       "X"s, and the values we want to generate here are "*"s:

               X   .   *   .   X

               .   .   .   .   .

               *   .   X   .   *

               .   .   .   .   .

               X   .   *   .   X

	       i and j represent our (x,y) position in the array. The
	       first value we want to generate is at (i=2,j=0), and we use
	       "oddline" and "stride" to increment j to the desired value.
	       */
		uint16_t oddline = 0;
		for (uint16_t x = 0; x < CELLEDGESIZE; x += stride)
		{
		    oddline = (oddline == 0);
			for (uint16_t y = 0; y < CELLEDGESIZE; y += stride)
			{
				if ((oddline) && !y)
				{
				    y += stride;
				}

			    if (!Seeded[x][y])
			    {
                    float Average = (Height[x - stride][y] + Height[x + stride][y] + Height[x][y - stride] + Height[x][y + stride]) * 0.25;
                    float RandomFactor = scale * RandGenerator->Roll(-1.0f, 1.0f);

                    Height[x][y] = Average + RandomFactor;

                    // Record Maximum and Minimum Height
                    if (Height[x][y] > CellTopZ)
                    {
                        CellTopZ = Height[x][y];
                    }
                    if (Height[x][y] < CellBottomZ)
                    {
                        CellBottomZ = Height[x][y];
                    }
			    }
				y += stride;
			}
		}

		/* reduce random number range. */
		scale *= ratio;
		stride >>= 1;
    }
}

int16_t Geology::getRockTypeAtCoordinates(CubeCoordinates Target, int32_t Zlevel)
{
    static uint16_t RockType0 = DATA->getLabelIndex("MATERIAL_BRIGHT_GRASS");
    static uint16_t RockType1 = DATA->getLabelIndex("MATERIAL_SNOW");
    static uint16_t RockType2 = DATA->getLabelIndex("MATERIAL_GRANITE");

    if (Zlevel > 8)
    {
        return RockType1;
    }
    if (Zlevel > 7)
    {
        return RockType2;
    }
    return RockType0;
}

TileShape Geology::getTileShapeAtCoordinates(CubeCoordinates CubeTarget, int32_t Zlevel)
{
    if (CubeTarget.X == 7 && CubeTarget.Y == 6 && Zlevel == 0)
    {
        bool Debug = true;
    }


    float Remainder;

    Remainder = Height[CubeTarget.X][CubeTarget.Y] - ((float) Zlevel);
    int8_t SWCornerHeight = min(max(roundf(Remainder * HEIGHT_FRACTIONS) + 1, 0.0f), (float) HEIGHT_FRACTIONS + 1);

    Remainder = Height[CubeTarget.X + 1][CubeTarget.Y] - ((float) Zlevel);
    int8_t SECornerHeight = min(max(roundf(Remainder * HEIGHT_FRACTIONS) + 1, 0.0f), (float) HEIGHT_FRACTIONS + 1);

    Remainder = Height[CubeTarget.X][CubeTarget.Y + 1] - ((float) Zlevel);
    int8_t NWCornerHeight = min(max(roundf(Remainder * HEIGHT_FRACTIONS) + 1, 0.0f), (float) HEIGHT_FRACTIONS + 1);

    Remainder = Height[CubeTarget.X + 1][CubeTarget.Y + 1] - ((float) Zlevel);
    int8_t NECornerHeight = min(max(roundf(Remainder * HEIGHT_FRACTIONS) + 1, 0.0f), (float) HEIGHT_FRACTIONS + 1);


    if (NECornerHeight == 0 || SWCornerHeight == 0)
    {
        return getTileShapeFromCornerHeight(SWCornerHeight, SECornerHeight, NWCornerHeight, NECornerHeight, true);
    }

    if (SECornerHeight == 0 || NWCornerHeight == 0)
    {
        return getTileShapeFromCornerHeight(SWCornerHeight, SECornerHeight, NWCornerHeight, NECornerHeight, false);
    }

    return getTileShapeFromCornerHeight(SWCornerHeight, SECornerHeight, NWCornerHeight, NECornerHeight, true);
}

void Geology::LoadCellData(Cell* TargetCell)
{
    CellCoordinates TargetCoordinates = TargetCell->getCellCoordinates();

    for (CubeCoordinates TargetCubeCoordinates; TargetCubeCoordinates.Index < (CUBESPERCELL - 1); ++TargetCubeCoordinates)
    {
        TileShape Shape = getTileShapeAtCoordinates(TargetCubeCoordinates, TargetCoordinates.Z);

        if (Shape != TILESHAPE_EMPTY)
        {
            int16_t MaterialType = getRockTypeAtCoordinates(TargetCubeCoordinates, TargetCoordinates.Z);
            TargetCell->setCubeMaterial(TargetCubeCoordinates, MaterialType);

            if (MaterialType != INVALID_INDEX)
            {
                TargetCell->setCubeShape(TargetCubeCoordinates, Shape);
                //TargetCell->setCubeSurface(TargetCubeCoordinates, RoughWallID);
            }
        }
        else
        {
            TargetCell->setCubeShape(TargetCubeCoordinates, Shape);
            TargetCell->setCubeMaterial(TargetCubeCoordinates, INVALID_INDEX);
        }
    }
}

File: src/Terrain/Geology.java
/* Copyright 2010 Kenneth 'Impaler' Ferland

 This file is part of Khazad.

 Khazad is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 Khazad is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Khazad.  If not, see <http://www.gnu.org/licenses/> */

package Terrain;

import Map.Coordinates.*;
import Core.Dice;
import Map.*;
import Data.DataManager;
import java.io.Serializable;

/**
 *
 * @author Impaler
 */
public class Geology implements Serializable {

	private static final long serialVersionUID = 1;

	int GeologySeed;
	Dice RandomGenerator;
	Noise NoiseGenerator;
	float[][] WorldHeight;

	short ChunkTopZ;
	short ChunkBottomZ;

	float[] Edge;
	float[][] Height;
	boolean[][] Seeded;

	public final short RockType0;
	public final short RockType1;
	public final short RockType2;

	public Geology() {
		DataManager Data = DataManager.getDataManager();

		RockType0 = Data.getLabelIndex("MATERIAL_SOIL");
		RockType1 = Data.getLabelIndex("MATERIAL_SNOW");
		RockType2 = Data.getLabelIndex("MATERIAL_GRANITE");
	}

	public boolean initialize(int Seed) {
		GeologySeed = Seed;

		RandomGenerator = new Dice();
		NoiseGenerator = new Noise();

		Height = new float[BlockCoordinate.CHUNK_EDGE_SIZE + 1][BlockCoordinate.CHUNK_EDGE_SIZE + 1];
		return true;
	}

	public short getChunkBottomZLevel() {
		return ChunkBottomZ;
	}

	public short getChunkTopZLevel() {
		return ChunkTopZ;
	}
	
	public void generateChunk(MapCoordinate targetCoordinates) {
		int X = targetCoordinates.getX() + GeologySeed;
		int Y = targetCoordinates.getY() + GeologySeed;

		int Octives = 4;
		double persistence = 0.4;

		for (int x = 0; x < BlockCoordinate.CHUNK_EDGE_SIZE + 1; x++) {
			for (int y = 0; y < BlockCoordinate.CHUNK_EDGE_SIZE + 1; y++) {

				double workingX = (x + X) / 200f;
				double workingY = (y + Y) / 200f;
				Height[x][y] = (float) OctavePerlin(workingX, workingY, Octives, persistence);
			}
		}
	}

	private double bias(double input, double bias) {
		return input + bias;
	}

	private double SmoothNoise(double x, double y) {
		double corners = ( NoiseGenerator.noise(x-1, y-1) + NoiseGenerator.noise(x+1, y-1) + NoiseGenerator.noise(x-1, y+1) + NoiseGenerator.noise(x+1, y+1) ) / 16;
		double sides   = ( NoiseGenerator.noise(x-1, y) + NoiseGenerator.noise(x+1, y) + NoiseGenerator.noise(x, y-1) + NoiseGenerator.noise(x, y+1) ) /  8;
		double center  =  NoiseGenerator.noise(x, y) / 4;

		return corners + sides + center;
	}

	public double OctavePerlin(double x, double y, int octaves, double persistence) {
		double total = 0;
		double raw;
		double frequency = 1;
		double amplitude = 20;

		for(int i = 0 ; i < octaves; i++) {
			raw = NoiseGenerator.noise(x * frequency, y * frequency);
			total += raw * amplitude;
			persistence *= raw + 1;
			amplitude *= persistence;
			frequency *= 2;
		}

		return total;
	}

	public void generateChunkEdge(float X, float Y, float heightScale, float Roughness) {
		ChunkCoordinate coords = new ChunkCoordinate((int) X, (int) Y, 0);
		coords.hashCode();
		RandomGenerator.seed(coords.hashCode());

		Edge[0] = X;
		Edge[BlockCoordinate.CHUNK_EDGE_SIZE] = Y;

		float ratio = (float) Math.pow(2.0, -Roughness);
		float scale = heightScale * ratio;

		/* Seed the endpoints of the array. To enable seamless wrapping,
		 * the endpoints need to be the same point. */
		short stride = BlockCoordinate.CHUNK_EDGE_SIZE / 2;

		while (stride > 0) {
			for (short i = stride; i < BlockCoordinate.CHUNK_EDGE_SIZE; i += stride) {
				Edge[i] = scale * RandomGenerator.roll(-1.0f, 1.0f) + ((Edge[i - stride] + Edge[i + stride]) * .5f);
				/* reduce random number range */
				scale *= ratio;
				i += stride;
			}
			stride >>= 1;
		}
	}

	public short getRockTypeAtCoordinates(short Target, int Zlevel) {

		if (Zlevel > 3) {
			return RockType1;
		}
		if (Zlevel > 1) {
			return RockType2;
		}
		return RockType0;
	}

	public BlockShape getBlockShapeAtCoordinates(int West, int East, int South, int North, int Bottom, int Top) {
		float Remainder;
		float BlockSize = Top - Bottom;

		Remainder = Height[West][South] - ((float) Bottom);
		Remainder = Math.round((Remainder / BlockSize) * ((float) BlockShape.HEIGHT_FRACTIONS)) + BlockShape.CUBE_BOTTOM_HEIGHT;
		byte SWCornerHeight = (byte) Math.min(Math.max(Remainder, BlockShape.BELOW_CUBE_HEIGHT), BlockShape.CUBE_TOP_HEIGHT);

		Remainder = Height[East][South] - ((float) Bottom);
		Remainder = Math.round((Remainder / BlockSize) * ((float) BlockShape.HEIGHT_FRACTIONS)) + BlockShape.CUBE_BOTTOM_HEIGHT;
		byte SECornerHeight = (byte) Math.min(Math.max(Remainder, (float) BlockShape.BELOW_CUBE_HEIGHT), (float) BlockShape.CUBE_TOP_HEIGHT);

		Remainder = Height[West][North] - ((float) Bottom);
		Remainder = Math.round((Remainder / BlockSize) * ((float) BlockShape.HEIGHT_FRACTIONS)) + BlockShape.CUBE_BOTTOM_HEIGHT;
		byte NWCornerHeight = (byte) Math.min(Math.max(Remainder, (float) BlockShape.BELOW_CUBE_HEIGHT), (float) BlockShape.CUBE_TOP_HEIGHT);

		Remainder = Height[East][North] - ((float) Bottom);
		Remainder = Math.round((Remainder / BlockSize) * ((float) BlockShape.HEIGHT_FRACTIONS)) + BlockShape.CUBE_BOTTOM_HEIGHT;
		byte NECornerHeight = (byte) Math.min(Math.max(Remainder, (float) BlockShape.BELOW_CUBE_HEIGHT), (float) BlockShape.CUBE_TOP_HEIGHT);

		byte Split = (byte) RandomGenerator.roll(0, 1);

		if (NWCornerHeight == BlockShape.CUBE_TOP_HEIGHT && SECornerHeight == BlockShape.CUBE_TOP_HEIGHT) {
			Split = 1;
		}

		if (SWCornerHeight == BlockShape.CUBE_TOP_HEIGHT && NECornerHeight == BlockShape.CUBE_TOP_HEIGHT) {
			Split = 0;
		}

		if (SWCornerHeight == BlockShape.BELOW_CUBE_HEIGHT || NECornerHeight == BlockShape.BELOW_CUBE_HEIGHT) {
			Split = 1;
		}

		if (SECornerHeight == BlockShape.BELOW_CUBE_HEIGHT || NWCornerHeight == BlockShape.BELOW_CUBE_HEIGHT) {
			Split = 0;
		}

		return new BlockShape(SWCornerHeight, SECornerHeight, NWCornerHeight, NECornerHeight, (byte) Split);
	}

	public void loadChunkData(Chunk TargetChunk) {
		ChunkCoordinate ChunkCoords = TargetChunk.getChunkCoordinates();
		for (byte DetailLevel = 0; DetailLevel < BlockCoordinate.CHUNK_DETAIL_LEVELS; DetailLevel++) {
			for (BlockCoordinate Index = new BlockCoordinate(DetailLevel); !Index.end(); Index.next()) {
				int Offset = 1 << DetailLevel;
				short x = (short) (Index.getX() * Offset);
				short y = (short) (Index.getY() * Offset);
				short z = (short) ((Index.getZ() * Offset) + (ChunkCoords.Z * BlockCoordinate.CHUNK_EDGE_SIZE));

				BlockShape Shape = getBlockShapeAtCoordinates(x, x + Offset, y, y + Offset, z, z + Offset);
				short MaterialType = getRockTypeAtCoordinates(Index.getBlockIndex(), z);
				if (MaterialType != DataManager.INVALID_INDEX) {
					TargetChunk.setBlockMaterial(Index, MaterialType);	
					TargetChunk.setBlockShape(Index, Shape);
				} else {
					TargetChunk.setBlockMaterial(Index, DataManager.INVALID_INDEX);
					TargetChunk.setBlockShape(Index, new BlockShape());
				}
			}
		}
	}
}
